pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2184,6 +2190,139 @@ public PinotResourceManagerResponse toggleInstance(String instanceName, boolean
+    int numReplicasInIdealState = Integer.parseInt(idealState.getReplicas());
+    final TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableNameWithType);
+    List<Map.Entry<String, Map<String,String>>> removedEntries = new LinkedList<>();
+    if (numReplicasInIdealState > targetNumReplicas) {
+      // We need to reduce the number of replicas per helix partition.
+      for (String segmentId : idealState.getPartitionSet()) {
+        Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentId);
+        if (instanceStateMap.size() > targetNumReplicas) {
+          Set<String> keys = instanceStateMap.keySet();
+          while (instanceStateMap.size() > targetNumReplicas) {
+            instanceStateMap.remove(keys.iterator().next());
+          }
+        } else if (instanceStateMap.size() < targetNumReplicas) {
+          LOGGER.warn("Table {}, segment {} has {} replicas, less than {} (requested number of replicas",
+              idealState.getResourceName(), segmentId, instanceStateMap.size(), targetNumReplicas);
+        }
+      }
+    } else {
+      final Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
+      if (tableType == TableType.REALTIME) {
+        removeSegmentsNotBalancable(mapFields, removedEntries);
+      }
+
+      // Number of replicas is either the same or higher, so invoke Helix rebalancer.
+      LinkedHashMap<String, Integer> states = new LinkedHashMap<>();
+      List<String> partitions = Lists.newArrayList(idealState.getPartitionSet());
+      states.put("OFFLINE", 0);
+      states.put("ONLINE", targetNumReplicas);
+      Set<String> currentHosts = new HashSet<>();
+      for (String segment : mapFields.keySet()) {
+        currentHosts.addAll(mapFields.get(segment).keySet());
+      }
+      AutoRebalanceStrategy rebalanceStrategy = new AutoRebalanceStrategy(tableNameWithType, partitions, states);
+
+      String serverTenant = TableNameBuilder.forType(tableType).tableNameWithType(tenantName);
+      List<String> instancesInClusterWithTag = _helixAdmin.getInstancesInClusterWithTag(_helixClusterName, serverTenant);
+      List<String> enabledInstancesWithTag =
+          HelixHelper.getEnabledInstancesWithTag(_helixAdmin, _helixClusterName, serverTenant);
+      LOGGER.info("Current nodes for table {}: {}", tableNameWithType, currentHosts);
+      LOGGER.info("New nodes for table {}: {}", tableNameWithType, instancesInClusterWithTag);
+      LOGGER.info("Enabled nodes for table: {}", tableNameWithType, enabledInstancesWithTag);
+      ZNRecord newZnRecord =
+          rebalanceStrategy.computePartitionAssignment(instancesInClusterWithTag, enabledInstancesWithTag,
+              mapFields, new ClusterDataCache());
+      final Map<String, Map<String, String>> newMapping = newZnRecord.getMapFields();
+      for (String segmentId : newMapping.keySet()) {
```

Comment:
for (Map.Entry<String, Map<String, String>> entry : newMapping.entrySet()) {
  idealState.setInstanceStateMap(entry.getKey(), entry.getValue());
}

Reply:
done